File: sdks/node/src/_i2i.ts
import * as fs from "node:fs";
import * as fsp from "node:fs/promises";
import * as path from "node:path";
import pLimit from "p-limit";
import { EnhancementClient } from "./_enhancement.js";
import { ProjectError, UploadError, ImagenError } from "./errors.js";
import { isValidImageFile } from "./utils.js";
import {
  unwrap,
  ProjectCreatedSchema,
  ProjectListResponseSchema,
  ProjectListItemSchema,
  PresignedFilesListSchema,
  TemporaryFileUploadDataSchema,
  MultipartUploadLinksResponseSchema,
  MessageResponseSchema,
  DownloadFilesListSchema,
  SingleDownloadLinkSchema,
  type ProjectListResponse,
  type ProjectListItem,
  type TemporaryFileUploadData,
  type MultipartUploadLinksResponse,
  type MessageResponse,
  type SingleDownloadLink,
  type I2IEditOptions,
  type UploadSummary,
  type UploadResult,
} from "./models.js";
import type { ProgressCallback } from "./_base.js";

/** S3 multipart part size: 64 MB (S3 requires >= 5 MB except for the final part). */
export const DEFAULT_MULTIPART_PART_SIZE = 64 * 1024 * 1024;

export interface ListI2iProjectsOptions {
  size?: number;
  page?: number;
  /** Filter by archived state, or null for all. Defaults to false (non-archived). */
  isArchived?: boolean | null;
}

export interface UploadI2iOptions {
  maxConcurrent?: number;
  calculateMd5?: boolean;
  onProgress?: ProgressCallback;
  /** Files larger than this many bytes use multipart upload. Defaults to 64 MB. */
  multipartThreshold?: number;
}

export interface MultipartUploadOptions {
  partSize?: number;
  maxConcurrent?: number;
}

/**
 * Image-to-Image (I2I) workflow (Workflow D): project create/list/validate/get,
 * size-routed upload (single presigned PUT or S3 multipart), start editing, and
 * download links. After `startI2iEditing`, wait for completion via a `callback_url`
 * or by calling `waitForI2iCompletion` (which polls the project status).
 */
export class I2IClient extends EnhancementClient {
  // -- Project lifecycle ----------------------------------------------------

  async createI2iProject(name?: string): Promise<string> {
    const body: Record<string, string> = {};
    if (name) body["name"] = name;

    this.logger.info(`Creating I2I project: ${name ?? "Unnamed"}`);
    const json = await this._makeRequest("POST", "/i2i/projects/", { json: body });
    const parsed = ProjectCreatedSchema.safeParse(unwrap(json));
    if (!parsed.success) {
      throw new ProjectError(`Could not parse I2I project creation response: ${parsed.error.message}`);
    }
    this.logger.info(`Created I2I project with UUID: ${parsed.data.projectUuid}`);
    return parsed.data.projectUuid;
  }

  async listI2iProjects(options: ListI2iProjectsOptions = {}): Promise<ProjectListResponse> {
    const { size = 20, page = 0, isArchived = false } = options;
    const params: Record<string, string | number | boolean> = { size, page };
    if (isArchived !== null) params["is_archived"] = isArchived;

    this.logger.debug(`Listing I2I projects (page=${page}, size=${size})`);
    const json = await this._makeRequest("GET", "/i2i/projects", { params });
    const parsed = ProjectListResponseSchema.safeParse(unwrap(json));
    if (!parsed.success) {
      throw new ProjectError(`Could not parse I2I project list response: ${parsed.error.message}`);
    }
    return parsed.data;
  }

  async validateI2iProjectName(name: string): Promise<boolean> {
    this.logger.debug(`Validating I2I project name: ${name}`);
    const json = await this._makeRequest("GET", "/i2i/projects/is_valid_name", {
      params: { name },
    });
    const data = unwrap(json);
    if (typeof data === "boolean") return data;
    if (data !== null && typeof data === "object") {
      for (const key of ["is_valid", "valid"]) {
        if (key in (data as Record<string, unknown>)) {
          return Boolean((data as Record<string, unknown>)[key]);
        }
      }
    }
    // A successful (2xx) response with no explicit flag is treated as valid.
    return true;
  }

  async getI2iProject(projectUuid: string, getThumbnail = true): Promise<ProjectListItem> {
    this.logger.debug(`Getting I2I project ${projectUuid}`);
    const json = await this._makeRequest("GET", `/i2i/projects/${projectUuid}`, {
      params: { get_thumbnail: getThumbnail },
    });
    const parsed = ProjectListItemSchema.safeParse(unwrap(json));
    if (!parsed.success) {
      throw new ProjectError(`Could not parse I2I project response: ${parsed.error.message}`);
    }
    return parsed.data;
  }

  // -- Uploads --------------------------------------------------------------

  /**
   * Upload images to an I2I project, routing each file by size: files at or below
   * `multipartThreshold` go through a single batched presigned PUT; larger files use
   * S3 multipart. Multipart is only available for I2I projects, which is why
   * auto-routing lives here rather than in the base `uploadImages`.
   */
  async uploadI2iImages(
    projectUuid: string,
    imagePaths: string[],
    options: UploadI2iOptions = {}
  ): Promise<UploadSummary> {
    const {
      maxConcurrent = 5,
      calculateMd5 = false,
      onProgress,
      multipartThreshold = DEFAULT_MULTIPART_PART_SIZE,
    } = options;
    if (maxConcurrent < 1) throw new Error("maxConcurrent must be at least 1");

    const validPaths = imagePaths.filter((p) => {
      try {
        return fs.statSync(p).isFile() && isValidImageFile(p);
      } catch {
        this.logger.warn(`Skipping invalid path: ${p}`);
        return false;
      }
    });
    if (validPaths.length === 0) {
      throw new UploadError("No valid local files found to upload.");
    }

    this.logger.info(`Starting I2I upload of ${validPaths.length} images to project ${projectUuid}`);

    // Route each file by size.
    const smallPaths: string[] = [];
    const largePaths: string[] = [];
    for (const p of validPaths) {
      if (fs.statSync(p).size > multipartThreshold) largePaths.push(p);
      else smallPaths.push(p);
    }

    const results: UploadResult[] = [];

    if (smallPaths.length > 0) {
      const filesList = smallPaths.map((filePath) => {
        const entry: Record<string, string> = { file_name: path.basename(filePath) };
        if (calculateMd5) entry["md5"] = this._calculateMd5(filePath);
        return entry;
      });

      const presignedJson = await this._makeRequest(
        "POST",
        `/i2i/projects/${projectUuid}/get_temporary_upload_links`,
        { json: { files_list: filesList, client_type: "API" } }
      );
      const parsed = PresignedFilesListSchema.safeParse(unwrap(presignedJson));
      if (!parsed.success) {
        throw new UploadError(`Could not parse I2I presigned URL response: ${parsed.error.message}`);
      }
      const uploadMap = new Map(parsed.data.files_list.map((f) => [f.fileName, f.uploadLink]));

      const limit = pLimit(maxConcurrent);
      const total = smallPaths.length;
      const batch = await Promise.all(
        smallPaths.map((filePath, index) =>
          limit(async (): Promise<UploadResult> => {
            onProgress?.(index, total, filePath);
            try {
              const uploadUrl = uploadMap.get(path.basename(filePath));
              if (!uploadUrl) {
                throw new UploadError(`No upload link found for ${path.basename(filePath)}`);
              }
              await this._uploadToS3(filePath, uploadUrl);
              this.logger.debug(`Uploaded: ${path.basename(filePath)}`);
              onProgress?.(index + 1, total, filePath);
              return { file: filePath, success: true, error: null };
            } catch (err) {
              this.logger.error(`Failed to upload ${path.basename(filePath)}: ${String(err)}`);
              onProgress?.(index + 1, total, filePath);
              return { file: filePath, success: false, error: String(err) };
            }
          })
        )
      );
      results.push(...batch);
    }

    for (const filePath of largePaths) {
      this.logger.info(`Routing large file to multipart upload: ${path.basename(filePath)}`);
      try {
        await this.uploadI2iFileMultipart(projectUuid, filePath, { maxConcurrent });
        results.push({ file: filePath, success: true, error: null });
      } catch (err) {
        this.logger.error(`Multipart upload failed for ${path.basename(filePath)}: ${String(err)}`);
        results.push({ file: filePath, success: false, error: String(err) });
      }
      onProgress?.(results.length, validPaths.length, filePath);
    }

    return {
      total: results.length,
      successful: results.filter((r) => r.success).length,
      failed: results.filter((r) => !r.success).length,
      results,
    };
  }

  async getI2iUploadLink(projectUuid: string, fileName: string): Promise<TemporaryFileUploadData> {
    this.logger.debug(`Getting I2I upload link for ${fileName} in ${projectUuid}`);
    const json = await this._makeRequest("GET", `/i2i/projects/${projectUuid}/get_upload_link`, {
      params: { file_name: fileName },
    });
    const parsed = TemporaryFileUploadDataSchema.safeParse(unwrap(json));
    if (!parsed.success) {
      throw new ProjectError(`Could not parse I2I upload link response: ${parsed.error.message}`);
    }
    return parsed.data;
  }

  // -- Multipart uploads ----------------------------------------------------

  async createI2iMultipartUpload(
    projectUuid: string,
    fileName: string,
    partCount: number
  ): Promise<MultipartUploadLinksResponse> {
    this.logger.debug(`Creating multipart upload for ${fileName} (${partCount} parts) in ${projectUuid}`);
    const json = await this._makeRequest("POST", `/i2i/projects/${projectUuid}/multipart_uploads`, {
      json: { file_name: fileName, part_count: partCount },
    });
    const parsed = MultipartUploadLinksResponseSchema.safeParse(unwrap(json));
    if (!parsed.success) {
      throw new UploadError(`Could not parse multipart upload links response: ${parsed.error.message}`);
    }
    return parsed.data;
  }

  async completeI2iMultipartUpload(
    projectUuid: string,
    uploadId: string,
    fileName: string
  ): Promise<void> {
    this.logger.debug(`Completing multipart upload ${uploadId} for ${fileName} in ${projectUuid}`);
    await this._makeRequest(
      "POST",
      `/i2i/projects/${projectUuid}/multipart_uploads/${uploadId}/complete`,
      { json: { file_name: fileName } }
    );
  }

  async abortI2iMultipartUpload(projectUuid: string, uploadId: string, key: string): Promise<void> {
    this.logger.debug(`Aborting multipart upload ${uploadId} in ${projectUuid}`);
    await this._makeRequest("DELETE", `/i2i/projects/${projectUuid}/multipart_uploads/${uploadId}`, {
      json: { key },
    });
  }

  /**
   * Upload a single large file using S3 multipart: splits the file into `partSize`
   * chunks, PUTs each part concurrently, then completes. On any failure the upload
   * is aborted before the error is re-raised. Peak memory is bounded to
   * `maxConcurrent * partSize` (each chunk is read inside the concurrency limit).
   */
  async uploadI2iFileMultipart(
    projectUuid: string,
    filePath: string,
    options: MultipartUploadOptions = {}
  ): Promise<MultipartUploadLinksResponse> {
    let { partSize = DEFAULT_MULTIPART_PART_SIZE } = options;
    const { maxConcurrent = 4 } = options;
    if (maxConcurrent < 1) throw new Error("maxConcurrent must be at least 1");

    let stat: fs.Stats;
    try {
      stat = fs.statSync(filePath);
      if (!stat.isFile()) throw new Error("not a file");
    } catch {
      throw new UploadError(`File not found for multipart upload: ${filePath}`);
    }

    const fileSize = stat.size;
    const fileName = path.basename(filePath);
    // S3 allows at most 10000 parts; grow part size if needed to stay within that.
    if (fileSize) partSize = Math.max(partSize, Math.ceil(fileSize / 10000));
    const partCount = Math.max(1, Math.ceil(fileSize / partSize));
    this.logger.info(`Multipart-uploading ${fileName} (${fileSize} bytes, ${partCount} parts) to ${projectUuid}`);

    const links = await this.createI2iMultipartUpload(projectUuid, fileName, partCount);

    const limit = pLimit(maxConcurrent);
    const fh = await fsp.open(filePath, "r");
    try {
      await Promise.all(
        links.parts.map((part) =>
          limit(async () => {
            const offset = (part.partNumber - 1) * partSize;
            const length = Math.max(0, Math.min(partSize, fileSize - offset));
            const buffer = Buffer.alloc(length);
            if (length > 0) await fh.read(buffer, 0, length, offset);
            const response = await fetch(part.uploadUrl, { method: "PUT", body: buffer });
            if (!response.ok) {
              throw new UploadError(`S3 part upload failed (${response.status}): ${response.statusText}`);
            }
            this.logger.debug(`Uploaded part ${part.partNumber}/${partCount} of ${fileName}`);
          })
        )
      );
      await this.completeI2iMultipartUpload(projectUuid, links.uploadId, fileName);
    } catch (err) {
      this.logger.error(`Multipart upload of ${fileName} failed: ${String(err)}; aborting upload ${links.uploadId}`);
      try {
        await this.abortI2iMultipartUpload(projectUuid, links.uploadId, links.key);
      } catch (abortErr) {
        this.logger.error(`Failed to abort multipart upload ${links.uploadId}: ${String(abortErr)}`);
      }
      throw new UploadError(`Multipart upload of ${fileName} failed: ${String(err)}`);
    } finally {
      await fh.close();
    }

    this.logger.info(`Completed multipart upload of ${fileName}`);
    return links;
  }

  // -- Editing & downloads --------------------------------------------------

  /**
   * Trigger I2I editing. This only triggers the edit and returns immediately;
   * to wait for completion, set a `callback_url` in `editOptions` or call
   * `waitForI2iCompletion`, which polls the project status until terminal.
   */
  async startI2iEditing(
    projectUuid: string,
    editOptions?: I2IEditOptions
  ): Promise<MessageResponse> {
    this.logger.info(`Triggering I2I edit for project ${projectUuid}`);
    const json = await this._makeRequest("POST", `/i2i/projects/${projectUuid}/edit`, {
      json: editOptions ?? {},
    });
    const parsed = MessageResponseSchema.safeParse(unwrap(json));
    if (!parsed.success) {
      throw new ImagenError(`Could not parse I2I edit trigger response: ${parsed.error.message}`);
    }
    return parsed.data;
  }

  /**
   * Poll an I2I project's status until editing completes or fails.
   *
   * Unlike the classic edit flow (which has a dedicated `/edit/status` endpoint),
   * I2I exposes its status on the project object: `GET /v1/i2i/projects/{uuid}`
   * returns a `status` of Pending -> In Progress -> Completed/Failed. This polls
   * that endpoint with exponential backoff, mirroring `_waitForCompletion`.
   *
   * @returns The final `ProjectListItem` once status is `Completed`.
   * @throws ProjectError if editing fails or the wait times out.
   */
  async waitForI2iCompletion(
    projectUuid: string,
    options: { pollIntervalMs?: number; maxWaitMs?: number } = {}
  ): Promise<ProjectListItem> {
    const { pollIntervalMs = 10_000, maxWaitMs = 72_000_000 } = options;
    const MAX_INTERVAL_MS = 60_000;
    let interval = pollIntervalMs;
    const startedAt = Date.now();

    this.logger.info(`Waiting for I2I editing of ${projectUuid} to complete...`);

    // eslint-disable-next-line no-constant-condition -- poll loop; exits on terminal status or timeout
    while (true) {
      if (Date.now() - startedAt > maxWaitMs) {
        throw new ProjectError(`I2I editing timed out after ${maxWaitMs / 1000}s`);
      }

      const project = await this.getI2iProject(projectUuid, false);
      const elapsed = Math.round((Date.now() - startedAt) / 1000);
      this.logger.info(`  I2I status: ${project.status} — elapsed ${elapsed}s`);

      if (project.status === "Completed") {
        this.logger.info("I2I editing completed successfully");
        return project;
      }

      if (project.status === "Failed") {
        throw new ProjectError(`I2I editing failed for project ${projectUuid}.`);
      }

      await new Promise<void>((resolve) => setTimeout(resolve, interval));
      interval = Math.min(interval * 1.2, MAX_INTERVAL_MS);
    }
  }

  async getI2iDownloadLinks(projectUuid: string): Promise<string[]> {
    this.logger.debug(`Getting I2I download links for project ${projectUuid}`);
    const json = await this._makeRequest(
      "GET",
      `/i2i/projects/${projectUuid}/get_temporary_download_links`
    );
    const parsed = DownloadFilesListSchema.safeParse(unwrap(json));
    if (!parsed.success) {
      throw new ProjectError(`Could not parse I2I download links response: ${parsed.error.message}`);
    }
    const links = parsed.data.files_list.map((f) => f.downloadLink);
    this.logger.info(`Retrieved ${links.length} I2I download links`);
    return links;
  }

  async getI2iDownloadLink(projectUuid: string, fileName: string): Promise<SingleDownloadLink> {
    this.logger.debug(`Getting I2I download link for ${fileName} in ${projectUuid}`);
    const json = await this._makeRequest("GET", `/i2i/projects/${projectUuid}/get_download_link`, {
      params: { file_name: fileName },
    });
    const parsed = SingleDownloadLinkSchema.safeParse(unwrap(json));
    if (!parsed.success) {
      throw new ImagenError(`Could not parse I2I download link response: ${parsed.error.message}`);
    }
    return parsed.data;
  }
}

File: sdks/go/i2i.go
package imagen

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"strconv"
	"sync"
	"time"
)

// I2IDefaultPartSize is the S3 multipart part size (and the default threshold
// above which a file is uploaded via multipart). S3 requires parts >= 5 MB
// except the final part.
const I2IDefaultPartSize = 64 << 20 // 64 MB

// maxS3Parts is S3's hard limit on parts per multipart upload.
const maxS3Parts = 10000

// CreateI2IProject creates an image-to-image project and returns its UUID.
func (c *Client) CreateI2IProject(ctx context.Context, name string) (string, error) {
	return c.createProject(ctx, "/i2i/projects/", name)
}

// ListI2IProjects returns a page of I2I projects. Only Size, Page and IsArchived
// are honoured for I2I.
func (c *Client) ListI2IProjects(ctx context.Context, opts *ListProjectsOptions) (*ProjectListResponse, error) {
	q := url.Values{}
	if opts != nil {
		if opts.Size != nil {
			q.Set("size", strconv.Itoa(*opts.Size))
		}
		if opts.Page != nil {
			q.Set("page", strconv.Itoa(*opts.Page))
		}
		if opts.IsArchived != nil {
			q.Set("is_archived", strconv.FormatBool(*opts.IsArchived))
		}
	}
	var out ProjectListResponse
	if err := c.doJSON(ctx, apiRequest{method: http.MethodGet, path: "/i2i/projects", query: q}, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// IsValidI2IName reports whether an I2I project name is available. A 2xx response
// with no explicit flag is treated as valid.
func (c *Client) IsValidI2IName(ctx context.Context, name string) (bool, error) {
	q := url.Values{}
	q.Set("name", name)
	raw, err := c.do(ctx, apiRequest{method: http.MethodGet, path: "/i2i/projects/is_valid_name", query: q})
	if err != nil {
		return false, err
	}
	var asBool bool
	if err := json.Unmarshal(raw, &asBool); err == nil {
		return asBool, nil
	}
	var obj struct {
		IsValid *bool `json:"is_valid"`
		Valid   *bool `json:"valid"`
	}
	if err := json.Unmarshal(raw, &obj); err == nil {
		if obj.IsValid != nil {
			return *obj.IsValid, nil
		}
		if obj.Valid != nil {
			return *obj.Valid, nil
		}
	}
	// 2xx with no explicit flag: treat as valid.
	return true, nil
}

// GetI2IProject fetches a single I2I project by UUID.
func (c *Client) GetI2IProject(ctx context.Context, projectUUID string, getThumbnail bool) (*ProjectListItem, error) {
	q := url.Values{}
	if getThumbnail {
		q.Set("get_thumbnail", "true")
	}
	var out ProjectListItem
	err := c.doJSON(ctx, apiRequest{
		method: http.MethodGet,
		path:   "/i2i/projects/" + url.PathEscape(projectUUID),
		query:  q,
	}, &out)
	if err != nil {
		return nil, err
	}
	return &out, nil
}

// GetI2IUploadLinks requests batched presigned PUT URLs for small files and
// returns a map of file name to upload URL.
func (c *Client) GetI2IUploadLinks(ctx context.Context, projectUUID string, files []FileUploadInfo) (map[string]string, error) {
	var out UploadLinksResponse
	err := c.doJSON(ctx, apiRequest{
		method: http.MethodPost,
		path:   fmt.Sprintf("/i2i/projects/%s/get_temporary_upload_links", url.PathEscape(projectUUID)),
		body:   uploadLinksRequest{FilesList: files, ClientType: "API"},
	}, &out)
	if err != nil {
		return nil, err
	}
	return uploadLinkMap(out), nil
}

// GetI2IUploadLink returns a single presigned PUT URL (advanced use).
func (c *Client) GetI2IUploadLink(ctx context.Context, projectUUID, fileName string) (string, error) {
	return c.singleLink(ctx, fmt.Sprintf("/i2i/projects/%s/get_upload_link", url.PathEscape(projectUUID)), fileName, true)
}

// CreateMultipartUpload starts an S3 multipart upload and returns per-part
// presigned URLs. partCount must be between 1 and 10000.
func (c *Client) CreateMultipartUpload(ctx context.Context, projectUUID, fileName string, partCount int) (*MultipartUploadResponse, error) {
	var out MultipartUploadResponse
	err := c.doJSON(ctx, apiRequest{
		method: http.MethodPost,
		path:   fmt.Sprintf("/i2i/projects/%s/multipart_uploads", url.PathEscape(projectUUID)),
		body:   map[string]any{"file_name": fileName, "part_count": partCount},
	}, &out)
	if err != nil {
		return nil, err
	}
	return &out, nil
}

// CompleteMultipartUpload finalizes a multipart upload.
func (c *Client) CompleteMultipartUpload(ctx context.Context, projectUUID, uploadID, fileName string) error {
	_, err := c.do(ctx, apiRequest{
		method: http.MethodPost,
		path:   fmt.Sprintf("/i2i/projects/%s/multipart_uploads/%s/complete", url.PathEscape(projectUUID), url.PathEscape(uploadID)),
		body:   map[string]string{"file_name": fileName},
	})
	return err
}

// AbortMultipartUpload cancels a multipart upload identified by its storage key.
func (c *Client) AbortMultipartUpload(ctx context.Context, projectUUID, uploadID, key string) error {
	_, err := c.do(ctx, apiRequest{
		method: http.MethodDelete,
		path:   fmt.Sprintf("/i2i/projects/%s/multipart_uploads/%s", url.PathEscape(projectUUID), url.PathEscape(uploadID)),
		body:   map[string]string{"key": key},
	})
	return err
}

// StartI2IEditing triggers I2I editing and returns immediately. Detect completion
// via the callback_url webhook or WaitForI2ICompletion (which polls the project's
// status field). opts may be nil.
func (c *Client) StartI2IEditing(ctx context.Context, projectUUID string, opts *I2IEditOptions) error {
	var body any
	if opts != nil {
		body = opts
	}
	_, err := c.do(ctx, apiRequest{
		method: http.MethodPost,
		path:   fmt.Sprintf("/i2i/projects/%s/edit", url.PathEscape(projectUUID)),
		body:   body,
	})
	return err
}

// GetI2IDownloadLinks returns all I2I result download links.
func (c *Client) GetI2IDownloadLinks(ctx context.Context, projectUUID string) (*DownloadLinksList, error) {
	return c.downloadLinks(ctx, fmt.Sprintf("/i2i/projects/%s/get_temporary_download_links", url.PathEscape(projectUUID)))
}

// GetI2IDownloadLink returns a single I2I result download link.
func (c *Client) GetI2IDownloadLink(ctx context.Context, projectUUID, fileName string) (string, error) {
	return c.singleLink(ctx, fmt.Sprintf("/i2i/projects/%s/get_download_link", url.PathEscape(projectUUID)), fileName, false)
}

// I2IUploadOptions tunes UploadI2IImages. The zero value uses the defaults.
type I2IUploadOptions struct {
	// CalculateMD5 computes and sends a base64 MD5 for the small-file batch.
	CalculateMD5 bool
	// MultipartThreshold is the size in bytes above which a file uses multipart
	// upload (default I2IDefaultPartSize, 64 MB).
	MultipartThreshold int64
	// PartSize is the bytes per multipart part (default I2IDefaultPartSize).
	PartSize int64
	// MaxConcurrency overrides the client default for both small-file uploads
	// and multipart part uploads (>= 1).
	MaxConcurrency int
	// Progress, if set, is called after each small file with (done, total, name).
	Progress func(done, total int, fileName string)
}

// UploadI2IImages uploads files to an I2I project, routing each by size: files at
// or below the threshold are batched into single PUTs, larger files use S3
// multipart upload. Files with an unsupported extension or that cannot be stat'd
// are silently skipped and do not appear in the summary (matching UploadImages);
// if no valid files remain, an error is returned.
func (c *Client) UploadI2IImages(ctx context.Context, projectUUID string, paths []string, opts *I2IUploadOptions) (*UploadSummary, error) {
	if opts == nil {
		opts = &I2IUploadOptions{}
	}
	threshold := opts.MultipartThreshold
	if threshold <= 0 {
		threshold = I2IDefaultPartSize
	}
	partSize := opts.PartSize
	if partSize <= 0 {
		partSize = I2IDefaultPartSize
	}
	concurrency := c.maxConcurrency
	if opts.MaxConcurrency >= 1 {
		concurrency = opts.MaxConcurrency
	}

	small, large, err := c.routeBySize(paths, threshold)
	if err != nil {
		return nil, err
	}
	if len(small) == 0 && len(large) == 0 {
		return nil, fmt.Errorf("imagen: no valid image files to upload")
	}
	if err := checkUniqueBaseNames(append(append([]string{}, small...), large...)); err != nil {
		return nil, err
	}

	var results []UploadResult

	if len(small) > 0 {
		infos, valid, err := prepareUploadInfos(small, opts.CalculateMD5)
		if err != nil {
			return nil, err
		}
		links, err := c.GetI2IUploadLinks(ctx, projectUUID, infos)
		if err != nil {
			return nil, err
		}
		results = append(results, c.runUploads(ctx, valid, links, md5ByName(infos), concurrency, opts.Progress).Results...)
	}

	for _, path := range large {
		name := baseName(path)
		if err := c.uploadFileMultipart(ctx, projectUUID, path, partSize, concurrency); err != nil {
			results = append(results, UploadResult{FileName: name, Err: fmt.Errorf("%w: %v", ErrUpload, err)})
		} else {
			results = append(results, UploadResult{FileName: name, Success: true})
		}
	}

	return summarize(results), nil
}

// routeBySize splits supported, existing files into small (<= threshold) and
// large (> threshold) buckets.
func (c *Client) routeBySize(paths []string, threshold int64) (small, large []string, err error) {
	for _, p := range paths {
		if !SupportedExtension(p) {
			continue
		}
		fi, statErr := os.Stat(p)
		if statErr != nil || !fi.Mode().IsRegular() {
			continue
		}
		if fi.Size() > threshold {
			large = append(large, p)
		} else {
			small = append(small, p)
		}
	}
	return small, large, nil
}

// uploadFileMultipart uploads one large file via S3 multipart upload, aborting
// the upload on failure.
func (c *Client) uploadFileMultipart(ctx context.Context, projectUUID, path string, partSize int64, concurrency int) error {
	fi, err := os.Stat(path)
	if err != nil {
		return err
	}
	fileSize := fi.Size()

	// S3 allows at most 10000 parts; grow partSize to stay within that. Use ceil
	// division so a file needing exactly 10001 parts is caught (integer
	// fileSize/partSize would round down and miss it).
	if ceilDiv(fileSize, partSize) > maxS3Parts {
		partSize = maxInt64(partSize, ceilDiv(fileSize, maxS3Parts))
	}
	partCount := int(maxInt64(1, ceilDiv(fileSize, partSize)))

	links, err := c.CreateMultipartUpload(ctx, projectUUID, baseName(path), partCount)
	if err != nil {
		return err
	}

	if err := c.uploadParts(ctx, path, partSize, links.Parts, concurrency); err != nil {
		c.abortMultipartBestEffort(projectUUID, links.UploadID, links.Key)
		return err
	}

	// Completing can also fail after every part succeeded; abort so a failed
	// complete does not leave the multipart upload dangling in S3.
	if err := c.CompleteMultipartUpload(ctx, projectUUID, links.UploadID, baseName(path)); err != nil {
		c.abortMultipartBestEffort(projectUUID, links.UploadID, links.Key)
		return err
	}
	return nil
}

// abortMultipartBestEffort tears down a multipart upload on a fresh context, so a
// canceled or expired caller context (often the very reason the upload failed)
// still lets the abort reach S3. Errors are ignored: the caller's original error
// is what matters.
func (c *Client) abortMultipartBestEffort(projectUUID, uploadID, key string) {
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()
	_ = c.AbortMultipartUpload(ctx, projectUUID, uploadID, key)
}

// uploadParts uploads each part concurrently. Chunks are read inside the
// semaphore so peak memory is bounded to concurrency * partSize.
func (c *Client) uploadParts(ctx context.Context, path string, partSize int64, parts []MultipartUploadPart, concurrency int) error {
	f, err := os.Open(path)
	if err != nil {
		return err
	}
	defer f.Close()

	sem := make(chan struct{}, concurrency)
	var wg sync.WaitGroup
	errs := make([]error, len(parts))

	for i, part := range parts {
		wg.Add(1)
		sem <- struct{}{}
		go func(i int, part MultipartUploadPart) {
			defer wg.Done()
			defer func() { <-sem }()
			offset := int64(part.PartNumber-1) * partSize
			chunk := make([]byte, partSize)
			n, rerr := f.ReadAt(chunk, offset)
			// ReadAt fills the buffer for a full part (nil error); only the final
			// part is short and returns io.EOF. A short read on any other part means
			// the file changed under us (e.g. truncation), so fail rather than
			// upload a corrupt, short part that would complete with bad data.
			isLast := part.PartNumber == len(parts)
			if rerr != nil && !(isLast && errors.Is(rerr, io.EOF) && n > 0) {
				errs[i] = fmt.Errorf("reading part %d: %w", part.PartNumber, rerr)
				return
			}
			if err := c.putBytes(ctx, part.UploadURL, chunk[:n]); err != nil {
				errs[i] = fmt.Errorf("uploading part %d: %w", part.PartNumber, err)
			}
		}(i, part)
	}
	wg.Wait()

	for _, e := range errs {
		if e != nil {
			return e
		}
	}
	return nil
}
